glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt

```diff
@@ -163,17 +177,49 @@ open class GleanInternalAPI internal constructor() {
         configuration: Configuration = Configuration(),
         buildInfo: BuildInfo
     ) {
-        // Glean initialization must be called on the main thread, or lifecycle
-        // registration may fail. This is also enforced at build time by the
-        // @MainThread decorator, but this run time check is also performed to
-        // be extra certain.
+        // Glean initialization must be called on the main thread for BOTH main and
+        // background processes.
+        //
+        // If we don't initialize on the main thread lifecycle registration may fail when
+        // initializing on the main process. This is also enforced at build time by the
+        // @MainThread decorator, but this run time check is also performed to be extra certain.
         ThreadUtils.assertOnUiThread()
 
-        // In certain situations Glean.initialize may be called from a process other than the main
-        // process.  In this case we want initialize to be a no-op and just return.
-        if (!isMainProcess(applicationContext)) {
-            Log.e(LOG_TAG, "Attempted to initialize Glean on a process other than the main process")
-            return
+        // If no `dataPath` is provided, then we setup Glean as usual.
+        if (configuration.dataPath == null) {
+            // In certain situations Glean.initialize may be called from a process other than
+            // the main process. In this case we want initialize to be a no-op and just return.
+            //
+            // You can call Glean.initialize from a background process, but to do so you need
+            // to specify a dataPath in the Glean configuration.
+            if (!isMainProcess(applicationContext)) {
+                Log.e(
+                    LOG_TAG,
+                    "Attempted to initialize Glean on a process other than the main process without a dataPath"
+                )
+                return
+            }
+
+            this.isCustomDataPath = false
+        } else {
+            // When the `dataPath` is provided, we need to make sure:
+            //   1. The database path provided is not `glean_data`.
+            //   2. The database path is valid and writable.
+
+            // The default database path is `glean_data`, the background process and the main
+            // process cannot write to the same file.
+            if (configuration.dataPath == "glean_data") {
+                Log.e(LOG_TAG, "Attempted to initialize Glean with an invalid database path \"glean_data\" is reserved")
+                return
+            }
+
+            // Check that the database path we are trying to write to is valid and writable.
+            if (!canWriteToDatabasePath(applicationContext.applicationInfo.dataDir, configuration.dataPath)) {
```

Comment:
I just realized that in the proposal we 

> MUST pass a valid database path in the app’s data directory (context.applicationInfo.dataDir).

I understood that as "the user passes in a full path".
this code currently only essentially takes a subpath and concatenates it to the data directory.

For Python we accept a full path (obviously, the concept of a fixed "data dir" only exists for mobile application as such).
Still I still feel like having the user pass in a full path is a better choice.
Or we need to rename that config option, cause what we want is more of a directory name then (at the very least this needs to be made clear in the docs)

@travis79 what do you think?

Reply:
I agree, explicitly passing in the full path makes it even harder for someone to misuse this.

Reply:
@badboy @travis79 

What constitutes the full path here? For example, when I print the path right now after I concatenate the custom slug, I get `/data/user/0/org.mozilla.samples.gleancore/glean_sample_data`.

So when we say full path, does that mean just from the app ID onward? So they would supply `org.mozilla.samples.gleancore/glean_sample_data`?